### PR TITLE
#160056373 Users should be able to recieve and use a reset password link

### PR DESCRIPTION
**Pull request by mppdg**
#### What does this PR do?
- This PR will enable a user to receive a reset password link and update his/her password.

#### Description of Task to be completed?
- provide forget password links to be added to`/ login` and `/signup` page
- create a page `/forgot-password` where a user can enter the email to receive the reset password link
- ensure user receives an email
- redirect a user to `/reset-password` page with the form to update a password.
- update a user password when he provides a valid password
- show the error messages when a password is not valid or not updated

#### How should this be manually tested?
- clone repo and install dependences with `npm install`
- `npm run dev` to run locally
- navigate to `/forgot-password` to enter an email and send link
- check email and click on the reset password link
- enter the new password and on the provided form

#### What are the relevant pivotal tracker stories?
#160056373

image
screen shot 2018-09-17 at 5 17 55 pm
screen shot 2018-09-17 at 5 17 04 pm
screen shot 2018-09-17 at 5 15 01 pm
screen shot 2018-09-17 at 5 14 28 pm
screen shot 2018-09-17 at 5 13 38 pm
screen shot 2018-09-17 at 5 12 46 pm

### CODE REVIEW

File: public/styles/App.scss [hunk 1]
```diff
@@ -0,0 +1,14 @@
+$body-color : green;
```
Comment (JCanaks) [inline public/styles/App.scss]: Pls refactor to follow the folder structure and convention styles

File: public/styles/ResetPassword.scss [hunk 2]
```diff
@@ -0,0 +1,67 @@
+@import url("https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css");
```
Comment (JCanaks) [inline public/styles/ResetPassword.scss]: Pls refactor to follow the folder structure and convention styles

File: src/views/user/ForgotPassword.js [hunk 3]
```diff
@@ -0,0 +1,120 @@
+import React from 'react';
```
Comment (JCanaks) [inline src/views/user/ForgotPassword.js]: Don't forget to covert all your component files to jsx

**Review by Fattylee — APPROVED** [review]

File: src/configs/routes.js [hunk 4]
```diff
@@ -12,104 +12,103 @@ const {
...
 const {
-  EditProfile, Profile, Followers, Following
+  EditProfile, Profile, Followers, Following,
+  ForgotPassword, ResetPassword,
 } = userPages;
 const { Signup, Login } = authPages;
 
 export default [
   {
-    id: 1,
```
Comment (megame24) [inline src/configs/routes.js]: Why did you remove the `id`s

File: src/services/requestHandler.js [hunk 5]
```diff
@@ -0,0 +1,37 @@
+import axios from 'axios';
+
+/**
+ * This function creates an instance of axios and return
+ * it for making API request
+ * @returns { instance } an instance of axios
+ */
+// const axiosInstance = () => {
```
Comment (megame24) [inline src/services/requestHandler.js]: No commented blocks of code

File: src/services/requestHandler.js [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+//   }
+//   const instanceCreate = axios.create({
+//     baseURL: apiUrl,
+//     headers: { 'Content-Type': 'application/json' },
+//   });
+//   return instanceCreate;
+// };
+
+
+const axiosInstance = (token = '') => {
```
Comment (megame24) [inline src/services/requestHandler.js]: There is an updated `axiosInstance`, you can rebase and make use of it.

File: src/views/user/ResetPassword/ResetPassword.jsx [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+  };
+
+  /**
+   * componentDidMoutn
+   * @returns {null} null
+   */
+  componentDidMount() {
+    const { location, dispatch, history } = this.props;
+    const resetToken = location.search.substr(7);
+    axiosInstance().get(`/users/reset-password?token=${resetToken}`)
```
Comment (megame24) [inline src/views/user/ResetPassword/ResetPassword.jsx]: You can make use of any of the middleware in place to handle this `async call`

**Comment by megame24** [comment]
Handle `PropTypes validation` in your tests

File: public/styles/ResetPassword.scss [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
```
Comment (rhenshaw56) [inline public/styles/ResetPassword.scss]: rename style file to have a kebab case, sorta like `reset-password.scss`

File: src/actions/resetPasswordActions.js [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+export const setToken = token => (
+  {
+    type: RESET_PASSWORD_TOKEN,
+    token,
+  }
+);
+
+const sendResetPasswordLink = (path, email) => {
+  dispatch(resetPasswordLink(path, email));
```
Comment (rhenshaw56) [inline src/actions/resetPasswordActions.js]: I had previously said not to use dispatch here since it seems sorta redundant to still call that sae action in side the same file

File: src/actions/resetPasswordActions.js [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+};
+
+const updatePassword = (path, password) => {
+  dispatch(resetPassword(path, password));
```
Comment (rhenshaw56) [inline src/actions/resetPasswordActions.js]: see comment above

File: src/actions/resetPasswordActions.js [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+const { RESET_PASSWORD_LINK, RESET_PASSWORD,
+  RESET_PASSWORD_TOKEN } = types;
+
+const resetPassword = (path, password) => (
+  {
+    type: RESET_PASSWORD,
+    payload: axiosInstance().put(
+      path,
+      { user: { password, confirm: password } },
```
Comment (rhenshaw56) [inline src/actions/resetPasswordActions.js]: I had previously stated to somehow combine reset password actions into the auth actions or user actions, same goes with the reducers

File: src/reducers/resetPasswordReducers.js [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+    status: false,
+    statusCode: 0,
+    message: '',
+  },
+};
+
+
+const resetPassword = (state = prevState, action = {}) => {
+  switch (action.type) {
+  case `${RESET_PASSWORD}_TOKEN`:
```
Comment (rhenshaw56) [inline src/reducers/resetPasswordReducers.js]: why don't you append the `_TOKEN` suffix in action types already since I know the promise middleware won't be doing that for you, except its a success, loading or failure

File: src/views/user/ResetPassword/ForgotPassword.jsx [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+    email: '',
+    emailErrors: [],
+  };
+
+  /**
+   * Renders the component
+   * @param {object} email event object
+   * @returns {undefined} returns nothing
+   */
+  validate = (email) => {
```
Comment (rhenshaw56) [inline src/views/user/ResetPassword/ForgotPassword.jsx]: I had previously said something about not using arrow functions in jsx context, pls change this and bind to the constructor if neccessary

File: src/views/user/ResetPassword/ForgotPassword.jsx [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+    const emailErrors = [];
+    if (!email) emailErrors.push('Email cannot be empty');
+
+    const regex = new RegExp(['^(([^<>()[\\]\\.,;:\\s@"]+(\\.',
```
Comment (rhenshaw56) [inline src/views/user/ResetPassword/ForgotPassword.jsx]: can you rename the regex variable name to be somewhat less generic ?

File: src/views/user/ResetPassword/ForgotPassword.jsx [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+    }
+    return true;
+  }
+
+  /**
+   * Renders the component
+   * @param {object} errorResonse error message object
+   * @returns {string} error message
+   */
+  serverErrorResponse = (errorResonse) => {
```
Comment (rhenshaw56) [inline src/views/user/ResetPassword/ForgotPassword.jsx]: please change all arrow functions to class methods

File: src/views/user/ResetPassword/ForgotPassword.jsx [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+          email={email}
+          success={success.status}
+        />
+      </div>
+    );
+  }
+}
+
+
+export const mapStateToProps = (state) => {
```
Comment (rhenshaw56) [inline src/views/user/ResetPassword/ForgotPassword.jsx]:
```
export const mapStateToProps = ({
  resetPasswordLink: { isLoading, success, error },
}) => ({
 isLoading,
 success,
 backendError: error
});
```

Comment (rhenshaw56) [reply src/views/user/ResetPassword/ForgotPassword.jsx]: what do you think ?

File: src/views/user/ResetPassword/ResetPassword.jsx [hunk 17]
```diff
@@ -0,0 +1,307 @@
...
```
Comment (rhenshaw56) [inline src/views/user/ResetPassword/ResetPassword.jsx]: I won't allow you to do this in the context of a react component, pls move this to an action to update state, that's what actions are for, cos if this call errors, how are we going to set an error state in redux store

File: src/views/user/ResetPassword/ResetPassword.jsx [hunk 18]
```diff
@@ -0,0 +1,307 @@
...
+   * @param {Object} input the target input
+   * @returns {undefined} no value
+   */
+  validatePassword = (input) => {
+    if (input.name !== 'password') return;
+    if (!input.value) {
+      this.validity.passwordIsValid = false;
+      return;
+    }
+    const regex = new RegExp(['^(?=.*[\\d])(?=.*[A-Z])(?=.*[a-z])',
```
Comment (rhenshaw56) [inline src/views/user/ResetPassword/ResetPassword.jsx]: renname regex

Comment (rhenshaw56) [reply src/views/user/ResetPassword/ResetPassword.jsx]: so I see you having plenty regex everywhere, do you think you can have a utils folder where you have helpers handling all regex cases, instead of doing that in a component

File: src/views/user/ResetPassword/ResetPassword.jsx [hunk 19]
```diff
@@ -0,0 +1,307 @@
...
+    });
+  }
+
+  /**
+   * handle password visibility
+   * @param {Object} event event object
+   * @returns {undefined} undefined
+   */
+  handlePasswordVisibility = (event) => {
+    const el = event.target;
```
Comment (rhenshaw56) [inline src/views/user/ResetPassword/ResetPassword.jsx]: what is `el` ?

**Review by rhenshaw56 — CHANGES_REQUESTED** [review]

**Review by megame24 — APPROVED** [review]

File: src/actions/resetPasswordActions.js (hunk 11 above)
Comment (mppdg) [reply src/actions/resetPasswordActions.js]: ok

File: src/views/user/ResetPassword/ForgotPassword.jsx (hunk 16 above)
Comment (mppdg) [reply src/views/user/ResetPassword/ForgotPassword.jsx]: It's robust!

File: src/views/user/ResetPassword/ForgotPassword.jsx (hunk 15 above)
Comment (mppdg) [reply src/views/user/ResetPassword/ForgotPassword.jsx]: I have changed all arrow functions to class methods except for this and two others because Eslint flagged that they must have a `this` call on them if changed to the normal method. But they don't have any reference to the `the` instance.

File: src/views/user/ResetPassword/ResetPassword.jsx (hunk 18 above)
Comment (mppdg) [reply src/views/user/ResetPassword/ResetPassword.jsx]: I used them only in two instances and both cases handle different patterns. So, they are not reused.

File: src/views/user/ResetPassword/ResetPassword.jsx (hunk 19 above)
Comment (mppdg) [reply src/views/user/ResetPassword/ResetPassword.jsx]: It's the target element. The element that triggered the event. I have renamed it to `target`.

File: src/actions/resetPasswordActions.js (hunk 9 above)
Comment (mppdg) [reply src/actions/resetPasswordActions.js]: I have made the adjustment

File: src/views/user/ResetPassword/ForgotPassword.jsx [hunk 20]
```diff
@@ -0,0 +1,164 @@
...
+    const { isLoading, success, backendError } = this.props;
+    const { emailErrors, email } = this.state;
+
+    const emailErrorMessage = emailErrors[0] ? emailErrors
+      .map((error, index) => <li key={index}>{error}</li>) : [''];
+
+    const serverResponse = { message: '', style: '' };
+
+    if (backendError.status) {
+      serverResponse.message = this.serverErrorResponse(backendError);
```
Comment (rhenshaw56) [inline src/views/user/ResetPassword/ForgotPassword.jsx]: `I have changed all arrow functions to class methods except for this and two others because Eslint flagged that they must have a this call on them if changed to the normal method. But they don't have any reference to the the instance` 

but now it's on the instance itself,
so either you move it to another file or you pull it out of the context of the component entirely

File: src/views/user/ResetPassword/ForgotPassword.jsx [hunk 21]
```diff
@@ -0,0 +1,164 @@
...
+  success: PropTypes.shape({
+    status: PropTypes.bool.isRequired,
+  }).isRequired,
+  backendError: PropTypes.shape({
+    status: PropTypes.bool.isRequired,
+  }).isRequired,
+  dispatch: PropTypes.func.isRequired,
+};
+
+export default connect(mapStateToProps, null)(ForgotPassword);
```
Comment (rhenshaw56) [inline src/views/user/ResetPassword/ForgotPassword.jsx]: what happens if there is no null ?

File: src/views/user/ResetPassword/ResetPassword.jsx [hunk 22]
```diff
@@ -0,0 +1,340 @@
...
+    }
+    this.validity.passwordIsValid = true;
+    this.validity.passwordIsConfirmed = confirm === input.value;
+  }
+
+  /**
+   * @param {Object} errorResonse http error status code
+   * @returns {String} error response message
+   */
+  serverErrorResponse = (errorResonse) => {
```
Comment (rhenshaw56) [inline src/views/user/ResetPassword/ResetPassword.jsx]: 👀

**Review by rhenshaw56 — CHANGES_REQUESTED** [review]

File: src/views/user/ResetPassword/ForgotPassword.jsx (hunk 20 above)
Comment (mppdg) [reply src/views/user/ResetPassword/ForgotPassword.jsx]: I have moved it to another file

File: src/views/user/ResetPassword/ForgotPassword.jsx (hunk 21 above)
Comment (mppdg) [reply src/views/user/ResetPassword/ForgotPassword.jsx]: It doesn't actually change anything. I thought it was required. I have removed it. Thanks

File: src/views/user/ResetPassword/ResetPassword.jsx (hunk 22 above)
Comment (mppdg) [reply src/views/user/ResetPassword/ResetPassword.jsx]: It's now been moved to a helper file

**Review by JCanaks — APPROVED** [review]

**Review by rhenshaw56 — APPROVED** [review]
